--do not merge-- Benchmark Branch for bytes-based Request

jackgopack4 opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
run benchmark for Persistent queue with following approach:
- request plus custom json-marshaled SpanContext in one storage op per item, added to buffer with bytes

<!-- Issue number if applicable -->
#### Link to tracking issue
Relates to https://github.com/open-telemetry/opentelemetry-collector/pull/12934

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Runs benchmark with persistent queue with filebacked storage extension

<!--Please delete paragraphs that you did not use before submitting.-->

jackgopack4 committed:
OTEL-2540 Add SpanContext to persistent queue

jackgopack4 committed:
OTEL-2540 add changelog

jackgopack4 committed:
OTEL-2540 add marshalRequestWithSpanContext and unmarshalRequestWithSpanContext

jackgopack4 committed:
add test coverage

mx-psi committed:
[chore] Update bugfix release guidelines (#12999)

<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

This updates the bugfix guidelines to make them less strict. In
particular, after this change, we would start releasing bugfix releases
under the following cases:

1. Lack of consensus of SIG leads on whether to release a bugfix version
within one working day after a report has been made
2. Issues that have not been reported by multiple people, but that are
known to be used in production

This also:
- Explicitly lists difficulties with debugging and troubleshooting as
'severe enough'
- Explicitly states that the release manager is responsible for bugfix
releases

jackgopack4 commented:
```
Running tool: /Users/john.peterson/.goenv/versions/1.23.8/bin/go test -benchmem -run=^$ -bench ^BenchmarkPersistentQueue_LargeRequests$ go.opentelemetry.io/collector/exporter/exporterhelper/internal/queuebatch

goos: darwin
goarch: arm64
pkg: go.opentelemetry.io/collector/exporter/exporterhelper/internal/queuebatch
cpu: Apple M3 Max
BenchmarkPersistentQueue_LargeRequests/original_request-16         	       1	11732299250 ns/op	5195660312 B/op	 4803913 allocs/op
BenchmarkPersistentQueue_LargeRequests/context_request-16          	       1	24771056667 ns/op	12312378384 B/op	 8902546 allocs/op
BenchmarkPersistentQueue_LargeRequests/context_and_key_value_request-16         	       1	11527782125 ns/op	5343508608 B/op	 9005812 allocs/op
PASS
ok  	go.opentelemetry.io/collector/exporter/exporterhelper/internal/queuebatch	49.773s
```

jackgopack4 commented:
similar results with different number of requests.
I specifically used exportertest from contrib to use an actual file-backed exporter to make sure the read/writes to/from file would be fast. I don't know if there's a way I can keep the file out of memory, and of course the Mac M3 have very fast memory and swap, anyway. Anyone is welcome to pull this branch and run the benchmark on their own machine to confirm.
But it looks like the double/triple unmarshaling of wrapping the request up with a SpanContext and marshaling again incurs much more overhead than just the added (double) storage operations.

jackgopack4 committed:
add persistentqueue benchmark test OTEL-2540

bogdandrutu committed:
[chore] Fix BenchmarkMemoryQueueWaitForResult test (#13006)

Fixes
http://github.com/open-telemetry/opentelemetry-collector/issues/13004

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

dmitryax commented on exporter/exporterhelper/internal/queuebatch/persistent_queue_benchmark_test.go:
```diff
@@ -0,0 +1,168 @@
+
+func (largeRequestEncoding) Unmarshal(buf []byte) (largeRequest, error) {
+	var req largeRequest
+	if err := json.Unmarshal(buf, &req); err != nil {
```
We should probably use one of the otlp signals with proto marshaling. I'll be much closer to reality than json. 

However, even with that, I guess it'll be pretty inefficient if we keep using these 3 marshal/unmarshal operations. I believe the main performance hit here is that we marshal/unmarshal every request batch twice (first by itself, and second with the context added). The whole chunk of memory allocated for the marshaled request is discarded right aways after being used for marshaling the result second time along with the context.

I think we need some custom marshaling/unmarshaling here instead of "json over json+proto". I couldn't give exact solution, but I would imagine we marshal context and then marshal the request right into the same bytes buffer so the buffer sequence would be something like {size of the context, context, request}

jackgopack4 replied:
Yes you're probably right. I will see if I can rework the logic to be more efficient marshaling

jackgopack4 replied:
OK, so I updated both the original PR and this branch to use a PTrace encoding for the request, as well as joining the payload via byte representation instead of a third marshal/unmarshal.

Even with more operations, the multiple storage operations are actually faster than the original method, as well as the method that was proposed first in the current PR.

```
john.peterson@COMP-HHVKJYJXCY queuebatch % go test -bench=BenchmarkPersistentQueue_PtraceTraces -benchmem -benchtime=5000000000x
goos: darwin
goarch: arm64
pkg: go.opentelemetry.io/collector/exporter/exporterhelper/internal/queuebatch
cpu: Apple M3 Max
BenchmarkPersistentQueue_PtraceTraces/original_request-16               5000000000               0.05113 ns/op         0 B/op         0 allocs/op
BenchmarkPersistentQueue_PtraceTraces/context_request-16                5000000000               0.06303 ns/op         1 B/op         0 allocs/op
BenchmarkPersistentQueue_PtraceTraces/context_and_key_value_request-16          5000000000               0.04285 ns/op       0 B/op           0 allocs/op
PASS
```

dmitryax replied:
That cannot be true. Adding another set of data to be handled separately cannot be faster than than plainly not doing it. Something must be wrong with the test suite. `0 allocs/op` is suspicious

jackgopack4 replied:
I'll take a closer look, thanks

jackgopack4 replied:
I see what the problem was; I was doing individual offer requests instead of batching like the existing benchmark. will make that change and report back.

mx-psi committed:
[chore] Add tests loading nil to any (#12998)

<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

While working on #12981, I would have found it useful to have these
tests, since they surfaced a bug when enabling `DecodeNil`.

mx-psi committed:
[chore] Make mapstructure hooks safe against untyped nils (#13001)

<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

If we enable `DecodeNil` as true, we may have [untyped
nils](https://go.dev/doc/faq#nil_error) being passed. Unfortunately,
these are not valid values for `reflect`, which leads to surprising
behavior such as golang/go/issues/51649.

Unfortunately, the default hooks from mapstructure do not deal with this
properly. To account for this, we:
- Vendor and change the `ComposeDecodeHookFunc` function so that this
case is accounted for the kinds of hooks that just won't work with
untyped nils
- Create a safe wrapper for the hooks that do work with untyped nils.
This wrapper is used in all hooks, but in the interest of keeping as
close to what I would imagine upstream will accept, I did not add this
to the compose function.

This should not have any end-user observable behavior.

<!-- Issue number if applicable -->
#### Link to tracking issue

Attempt to work around
https://github.com/open-telemetry/opentelemetry-collector/pull/12996#issuecomment-2862859367

---------

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

jackgopack4 committed:
switch marshal from JSON to byte-based

dmitryax commented on exporter/exporterhelper/internal/queuebatch/persistent_queue.go:
```diff
@@ -313,20 +476,49 @@ func (pq *persistentQueue[T]) Read(ctx context.Context) (context.Context, T, Don
+	ops = append(ops, storage.SetOperation(readIndexKey, itemIndexToBytes(pq.readIndex)))
+	ops = append(ops, storage.SetOperation(currentlyDispatchedItemsKey, itemIndexArrayToBytes(pq.currentlyDispatchedItems)))
+	ops = append(ops, getOp)
+	if ctx.Value(requestTypeKey) != nil {
```
Passing the benchmark test options as part of the context might be convenient, but I guess that can make the benchmark results incorrect since that logic adds its own cost. If you could just split the solutions in two PRs and keep the same benchmark test that would be great. The benchmark test can even go to another PR and be merged upstream first

jackgopack4 replied:
yes, I've done this and will open two separate PRs

jackgopack4 replied:
the problem with the new benchmark is I am using a contrib component (extension/storage/storagetest) and that gives otelcol a bunch of new dependencies. There may be a way to migrate some of the storagetest code to collector core and then import it in contrib

jackgopack4 committed:
switch approach to make multiple storage operations OTEL-2540

jackgopack4 committed:
add new benchmark with filebacked storage extension OTEL-2540

jackgopack4 committed:
remove inadvertent debug code

jackgopack4 committed:
fix new benchark

jackgopack4 committed:
Revert "switch approach to make multiple storage operations OTEL-2540"

This reverts commit c49367e2fcc894d17323d89d5ff9d8e1e5e1a37c.

jackgopack4 commented:
```
go test -benchmem -run=^$ -bench ^BenchmarkPersistentQueue_PtraceTraces$ go.opentelemetry.io/collector/exporter/exporterhelper/internal/queuebatch

goos: darwin
goarch: arm64
pkg: go.opentelemetry.io/collector/exporter/exporterhelper/internal/queuebatch
cpu: Apple M3 Max
BenchmarkPersistentQueue_PtraceTraces-16    	    2533	    541251 ns/op	 6210329 B/op	     639 allocs/op
PASS
ok  	go.opentelemetry.io/collector/exporter/exporterhelper/internal/queuebatch	2.055s
```

jackgopack4 committed:
make gotidy

dmitryax commented on exporter/exporterhelper/internal/queuebatch/persistent_queue.go:
```diff
@@ -238,6 +244,140 @@ func (pq *persistentQueue[T]) Offer(ctx context.Context, req T) error {
+	buf := make([]byte, 0, 8+len(reqBuf)+len(scJSON))
+	//nolint:gosec // G115: integer overflow conversion int -> uint32
+	buf = binary.LittleEndian.AppendUint32(buf, uint32(len(reqBuf)))
+	buf = append(buf, reqBuf...)
```
Ok this still requires whole byte array reallocation which is a huge performance hit. I was under impression that we can marshal request into a preallocated array, but apparently pdata doesn't expose that interface. I see there is a another request for that https://github.com/open-telemetry/opentelemetry-collector/issues/12395. Until that capability is introduced, we can go with the two separate keys under alpha feature gate. Once https://github.com/open-telemetry/opentelemetry-collector/issues/12395 is resolved, we should be able to get back to this approach with performance improvement.

dmitryax replied:
But before jumping on https://github.com/open-telemetry/opentelemetry-collector/issues/12395, we should probably export and use https://github.com/open-telemetry/opentelemetry-collector/blob/main/pdata/internal/data/protogen/trace/v1/trace.pb.go#L951 in this PR to see the benchmark numbers that we would get eventually. If you could do that whenever you have time, that would be perfect. No need to block the work on https://github.com/open-telemetry/opentelemetry-collector/pull/12934 with two separate keys under an alpha feature gate.
